### PR TITLE
Memory > Diff: UX improvements.

### PR DESCRIPTION
**Pull request by polina-c**
1. Move class filter from control pane to column header to make it discoverable (see session 5 in go/devtools-memory-diff-ux-research)

2. Update diff guidance to require understanding of memory concepts

3. Move word 'snapshot' from documentation item to tab header

4. Add noun to each tab header to make them more verbose:
    - before: Profile, Diff, Trace
    - after: Profile Memory, Diff Snapshots, Trace Instances

https://user-images.githubusercontent.com/12115586/211207985-785592dd-0070-4a58-9028-ac5348aff464.mov

Screenshot 2023-01-08 at 8 26 27 AM

### CODE REVIEW

**Comment by polina-c** [comment]
@CoderDake 

@melyndahoover

File: packages/devtools_app/lib/src/screens/memory/panes/diff/diff_pane.dart [hunk 1]
```diff
@@ -107,22 +114,34 @@ class SnapshotInstanceItemPane extends StatelessWidget {
...
-Take a **heap snapshot** to view current memory allocation:
+1. Understand [Dart memory concepts](https://docs.flutter.dev/development/tools/devtools/memory#basic-memory-concepts).
+
+\v
+
+2. Take a **heap snapshot** to view current memory allocation:
+
+    a. In the Snapshots panel, click the ● button
+
+    b. Use the **Filter** button to refine the results
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/diff_pane.dart:125]: Maybe we should add an "optional" tag to this instruction: `b. [Optional] use the **Filter** button to refine the results`

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/classes_table_diff.dart [hunk 2]
```diff
@@ -190,31 +232,25 @@ class ClassesTableDiff extends StatelessWidget {
...
-    _retainedSizeDeltaColumn,
-  ];
+  static final _columnStore = <String, _ClassesTableDiffColumns>{};
+  static _ClassesTableDiffColumns _columns(
+    Widget classFilterButton,
+  ) =>
+      _columnStore.putIfAbsent(
+        '${identityHashCode(classFilterButton)}',
+        () => _ClassesTableDiffColumns(classFilterButton),
+      );
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/classes_table_diff.dart:244]: this is confusing. Why do we have a column store where we store a set of columns based on a widget's hash code?

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart [hunk 3]
```diff
@@ -125,12 +126,10 @@ class _SnapshotListTitle extends StatelessWidget {
...
           if (theItem is SnapshotDocItem)
-            Expanded(
-              child: Text(
-                'Snapshots',
-                overflow: TextOverflow.ellipsis,
-                style: textStyle,
-              ),
+            Icon(
+              Icons.help_outline,
+              size: questionMarkSize,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/snapshot_list.dart:131]: Can we use `defaultIconSize` here to keep this consistent with other icon buttons throughout DevTools?

File: packages/devtools_app/test/memory/allocation_tracing/allocation_profile_tracing_view_test.dart [hunk 4]
```diff
@@ -151,8 +151,7 @@ void main() {
         final refresh = find.text('Refresh');
         expect(refresh, findsOneWidget);
 
-        // Tab name and column name.
-        expect(find.text('Trace'), findsNWidgets(2));
+        expect(find.text('Trace'), findsNWidgets(1));
```
Comment (kenzieschmoll) [inline packages/devtools_app/test/memory/allocation_tracing/allocation_profile_tracing_view_test.dart:154]: replace with findsOneWidget instead

File: tool/release_notes/NEXT_RELEASE_NOTES.md [hunk 5]
```diff
@@ -20,7 +20,8 @@ TODO: Remove this section if there are not any general updates.
 TODO: Remove this section if there are not any general updates.
 
 ## Memory updates
-TODO: Remove this section if there are not any general updates.
+
+- Improve useability of snapshot diffing [#5015](https://github.com/flutter/devtools/pull/5015)
```
Comment (CoderDake) [inline tool/release_notes/NEXT_RELEASE_NOTES.md:24]:
```suggestion
- Improve usability of snapshot diffing [#5015](https://github.com/flutter/devtools/pull/5015)
```

**Review by CoderDake — APPROVED** [review]

File: packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/classes_table_diff.dart (hunk 2 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/screens/memory/panes/diff/widgets/classes_table_diff.dart:244]: replaced with snapshotId

**Pull request merged**